File: discipline_ts/core/vault.ts
import { VaultName, VaultData, Countdown, Instant, Duration, Unique } from "../x.ts";

const BRAND = Symbol();

type RawVault = {
  name: VaultName,
  data: VaultData,
  protection: Countdown,
};

export type Vault = Unique<typeof BRAND, "Vault", RawVault>;

const construct = (
  name: VaultName,
  data: VaultData,
  protection: Countdown,
): Vault => {
  return {
    name,
    data,
    protection,
  } satisfies RawVault as Vault;
};

const reconstruct = construct;

const create = construct;

const getName = (it: Vault): VaultName => {
  return it.name;
};

const getData = (it: Vault): VaultData => {
  return it.data;
};

const getProtection = (it: Vault): Countdown => {
  return it.protection;
};

const isProtected = (it: Vault, now: Instant): boolean => {
  return Countdown.isRunning(it.protection, now);
};

const extendProtectionByOrSetToMaxSafeValue = (
  it: Vault,
  now: Instant,
  factor: Duration,
): void => {
  if (Countdown.isFinished(it.protection, now)) {
    it.protection = Countdown.create(
      now,
      Duration.min(MAXIMUM_PROTECTION_DURATION, factor),
    );
  } else {
    const remaining = Countdown.getRemainingTimeOrZero(it.protection, now);
    const maximum = MAXIMUM_PROTECTION_DURATION;

    if (Duration.isLongerThan(Duration.saturatingAdd(factor, remaining), maximum)) {
      factor = Duration.saturatingSub(maximum, remaining);
    }

    Countdown.saturatingIncrementDuration(it.protection, factor);
  }
};

export const Vault = {
  reconstruct,
  create,
  construct,
  getName,
  getData,
  getProtection,
  isProtected,
  extendProtectionByOrSetToMaxSafeValue,
};

File: discipline_kt/app/src/main/java/com/example/app/vaults/Vault.kt
package com.example.app

public data class Vault private constructor(
  val name: VaultName,
  val data: VaultData,
  var protector: VaultProtector
) {
  companion object {
    // val MAXIMUM_PROTECTION_DURATION = Duration.fromDays(7).getOrThrow()
    
    fun create(
      name: VaultName,
      data: VaultData,
      protector: VaultProtector,
    ): Vault {
      // if (protection.getTotalDuration().isLongerThan(MAXIMUM_PROTECTION_DURATION)) {
      //   return Tried.failure(
      //     TextualError.create("Creating a Vault")
      //       .addMessage("Vault protection countdown's total duration is longer than the maximum protection duration")
      //       .addStringAttachment("Protection countdown duration", protection.getTotalDuration().toString())
      //       .addStringAttachment("Maximum protection duration", MAXIMUM_PROTECTION_DURATION.toString())
      //   )
      // }

      return Vault(name, data, protector)
    }

    // fun createOrThrow(
    //   name: VaultName,
    //   data: VaultData,
    //   protection: Countdown
    // ): Vault {
    //   if (protection.getTotalDuration().isLongerThan(MAXIMUM_PROTECTION_DURATION)) {
    //     throw TextualError.create("Creating a Vault")
    //       .addMessage("Vault protection countdown's total duration is longer than the maximum protection duration")
    //       .addStringAttachment("Protection countdown duration", protection.getTotalDuration().toString())
    //       .addStringAttachment("Maximum protection duration", MAXIMUM_PROTECTION_DURATION.toString())
    //   }

    //   return Vault(name, data, protection)
    // }
    
    // /**
    //  * Creates a Vault with validation
    //  */
    // fun createValidated(
    //   name: String,
    //   data: String,
    //   protectionDuration: Duration,
    //   protectionStart: Instant
    // ): Either<TextualError, Vault> = either {
    //   val vaultName = VaultName.create(name).bind()
    //   val vaultData = VaultData.create(data).bind()
      
    //   if (protectionDuration.isLongerThan(MAXIMUM_PROTECTION_DURATION)) {
    //     raise(
    //       TextualError.create("Creating Vault")
    //         .addMessage("Protection duration exceeds maximum")
    //         .addStringAttachment("Maximum duration", MAXIMUM_PROTECTION_DURATION.toString())
    //         .addStringAttachment("Provided duration", protectionDuration.toString())
    //     )
    //   }
      
    //   val countdown = Countdown.create(protectionStart, protectionDuration)
    //   Vault(vaultName, vaultData, countdown)
    // }
  }

  fun isProtected(now: Instant): Boolean {
    return protector.isRunning(now)
  }
  
  // /**
  //  * Extends the protection duration safely, respecting maximum limits
  //  */
  // // TODO: Check whether this is implemented correctly
  // fun extendProtectionByOrSetToMaxSafeValue(now: Instant, factor: Duration) {
  //   if (protection.isFinished(now)) {
  //     val newDuration = MAXIMUM_PROTECTION_DURATION.min(factor)
  //     val newProtection = Countdown.create(now, newDuration)
  //     protection = newProtection
  //     return
  //   } 
    
  //   val remaining = protection.getRemainingTimeOrZero(now)
  //   val maximum = MAXIMUM_PROTECTION_DURATION
    
  //   var actualFactor = factor
  //   if (actualFactor.saturatingAdd(remaining).isLongerThan(maximum)) {
  //     actualFactor = maximum.saturatingSub(remaining)
  //   }
    
  //   val newProtection = protection.copy(duration = protection.getTotalDuration().saturatingAdd(actualFactor))
  //   protection.set = newProtection
  // }

}
